### PR TITLE
[Clang][ARM] Call constructor on BranchTargetInfo.

### PR DESCRIPTION
**Pull request by DanielKristofKiss**
Otherwise members will be uninitialised.

### CODE REVIEW

**Review by jthackray — APPROVED** [review]

**Comment by tmatheson-arm** [comment]
Can something be done to prevent this happening accidentally again?

**Comment by jthackray** [comment]
> Can something be done to prevent this happening accidentally again?

+1. Currently, the struct is initialised with:
`    BranchProtectionInfo() = default;`
so it appears this could be improved in future.

**Comment by s-barannikov** [comment]
IIRC there is a clang-tidy check that diagnoses uninitialized PODs, but probably making the constructor non-trivial is simpler.

**Review by s-barannikov — APPROVED** [review]

File: clang/include/clang/Basic/TargetInfo.h [hunk 1]
```diff
@@ -1434,7 +1434,14 @@ class TargetInfo : public TransferrableTargetInfo,
...
 
   public:
-    BranchProtectionInfo() = default;
+    BranchProtectionInfo() {
+      SignReturnAddr = LangOptions::SignReturnAddressScopeKind::None;
+      SignKey = LangOptions::SignReturnAddressKeyKind::AKey;
+      BranchTargetEnforcement = false;
+      BranchProtectionPAuthLR = false;
+      GuardedControlStack = false;
+    };
```
Comment (tmatheson-arm) [inline clang/include/clang/Basic/TargetInfo.h:1443]:
```suggestion
    BranchProtectionInfo() :
      SignReturnAddr(LangOptions::SignReturnAddressScopeKind::None),
      SignKey(LangOptions::SignReturnAddressKeyKind::AKey),
      BranchTargetEnforcement(false),
      BranchProtectionPAuthLR(false),
      GuardedControlStack(false)
    {}
```
This is more likely to show up warnings if something gets missed or reordered.

File: clang/lib/CodeGen/Targets/ARM.cpp [hunk 2]
```diff
@@ -141,7 +141,7 @@ class ARMTargetCodeGenInfo : public TargetCodeGenInfo {
       ParsedTargetAttr Attr =
           CGM.getTarget().parseTargetAttr(TA->getFeaturesStr());
       if (!Attr.BranchProtection.empty()) {
-        TargetInfo::BranchProtectionInfo BPI;
+        TargetInfo::BranchProtectionInfo BPI{};
```
Comment (tmatheson-arm) [inline clang/lib/CodeGen/Targets/ARM.cpp:144]: These are unnecessary now, right? Making them explicit is fine though.

Comment (DanielKristofKiss) [reply clang/lib/CodeGen/Targets/ARM.cpp:144]: Right, now unnecessary, maybe more stylish.

File: clang/include/clang/Basic/TargetInfo.h (hunk 1 above)
Comment (DanielKristofKiss) [reply clang/include/clang/Basic/TargetInfo.h]: right, added and clang-formated.

**Review by tommat01 — APPROVED** [review]

**Pull request merged**